modules/translation/i18n/i18n.go

```diff
@@ -25,174 +25,273 @@ var (
+				idx = len(store.trKeyToIdxMap)
+				store.trKeyToIdxMap[trKey] = idx
 			}
-			lc.textMap[textIdx] = key.Value()
+			idxToMsgMap[idx] = key.Value()
 		}
 	}
 	iniFile = nil
-	return nil
+	return idxToMsgMap, nil
+}
+
+func (store *LocaleStore) idxForTrKey(trKey string) (int, bool) {
+	if store.reloadMu != nil {
+		store.reloadMu.RLock()
+		defer store.reloadMu.RUnlock()
+	}
+	idx, ok := store.trKeyToIdxMap[trKey]
+	return idx, ok
 }
 
-func (ls *LocaleStore) HasLang(langName string) bool {
-	_, ok := ls.localeMap[langName]
+// HasLang reports if a language is available in the store
+func (store *LocaleStore) HasLang(langName string) bool {
+	_, ok := store.localeMap[langName]
 	return ok
 }
 
-func (ls *LocaleStore) ListLangNameDesc() (names, desc []string) {
-	return ls.langNames, ls.langDescs
+// ListLangNameDesc reports if a language available in the store
+func (store *LocaleStore) ListLangNameDesc() (names, desc []string) {
+	return store.langNames, store.langDescs
 }
 
 // SetDefaultLang sets default language as a fallback
-func (ls *LocaleStore) SetDefaultLang(lang string) {
-	ls.defaultLang = lang
+func (store *LocaleStore) SetDefaultLang(lang string) {
+	store.defaultLang = lang
 }
 
 // Tr translates content to target language. fall back to default language.
-func (ls *LocaleStore) Tr(lang, trKey string, trArgs ...interface{}) string {
-	l, ok := ls.localeMap[lang]
+func (store *LocaleStore) Tr(lang, trKey string, trArgs ...interface{}) string {
+	l, ok := store.localeMap[lang]
 	if !ok {
-		l, ok = ls.localeMap[ls.defaultLang]
+		l, ok = store.localeMap[store.defaultLang]
 	}
+
 	if ok {
 		return l.Tr(trKey, trArgs...)
 	}
 	return trKey
 }
 
+// reloadIfNeeded will check if the locale needs to be reloaded
+// this function will assume that the l.reloadMu has been RLocked if it already exists
+func (l *locale) reloadIfNeeded() {
+	if l.reloadMu == nil {
+		return
+	}
+
+	now := time.Now()
+	if now.Sub(l.lastReloadCheckTime) < time.Second || l.sourceFileInfo == nil || l.sourceFileName == "" {
+		return
+	}
+
+	l.reloadMu.RUnlock()
+	l.reloadMu.Lock() // (NOTE: a pre-emption can occur between these two locks so we need to recheck)
+
+	if now.Sub(l.lastReloadCheckTime) < time.Second || l.sourceFileInfo == nil || l.sourceFileName == "" {
+		l.reloadMu.Unlock()
+		l.reloadMu.RLock()
+		return
+	}
+
+	l.lastReloadCheckTime = now
+	sourceFileInfo, err := os.Stat(l.sourceFileName)
+	if err != nil || sourceFileInfo.ModTime().Equal(l.sourceFileInfo.ModTime()) {
+		l.reloadMu.Unlock()
+		l.reloadMu.RLock()
+		return
+	}
+
+	idxToMsgMap, err := l.store.readIniToIdxToMsgMap(l.sourceFileName)
+	if err == nil {
+		l.idxToMsgMap = idxToMsgMap
+	} else {
+		log.Error("Unable to live-reload the locale file %q, err: %v", l.sourceFileName, err)
+	}
+
+	// We will set the sourceFileInfo to this file to prevent repeated attempts to re-load this broken file
+	l.sourceFileInfo = sourceFileInfo
+
+	l.reloadMu.Unlock()
+	l.reloadMu.RLock()
+}
```

Comment:
```suggestion
	l.reloadMu.RUnlock()
	l.reloadMu.Lock() // (NOTE: a pre-emption can occur between these two locks so we need to recheck)
	defer l.reloadMu.RLock()
	defer l.reloadMu.Unlock()

	if now.Sub(l.lastReloadCheckTime) < time.Second || l.sourceFileInfo == nil || l.sourceFileName == "" {
		return
	}

	l.lastReloadCheckTime = now
	sourceFileInfo, err := os.Stat(l.sourceFileName)
	if err != nil || sourceFileInfo.ModTime().Equal(l.sourceFileInfo.ModTime()) {
		return
	}

	idxToMsgMap, err := l.store.readIniToIdxToMsgMap(l.sourceFileName)
	if err == nil {
		l.idxToMsgMap = idxToMsgMap
	} else {
		log.Error("Unable to live-reload the locale file %q, err: %v", l.sourceFileName, err)
	}

	// We will set the sourceFileInfo to this file to prevent repeated attempts to re-load this broken file
	l.sourceFileInfo = sourceFileInfo
}
```